### PR TITLE
Upgrade openamp and libmetal to last commit

### PR DESCRIPTION
**Pull request by CV-Bowen**
## Summary
-    virtio-mmio/pci: add alloc_buf/free_buf for mmio and pci transport layer
    
    And remove the virtio_alloc/free_buf implementation in virtio.c, because
    these two apis has been moved to the OpenAMP

-    rptun: rptun implements notify_wait through rvdev->notify_wait_cb
    
    notify_wait has been remove in remoteproc ops and virtio_dispatch and
    has been moved to the struct rpmsg_virtio_device in new OpenAMP,
    so change the rptun too,

-    virtio_9p.c: fix compile error in OpenAMP upgrade feature reflow to dev
    
    v9fs/virtio_9p.c:223:3: error: too few arguments to function 'virtio_negotiate_features'
      223 |   virtio_negotiate_features(vdev, 1 << VIRTIO_9P_MOUNT_TAG);
    v9fs/virtio_9p.c:245:9: error: too few arguments to function 'virtio_create_virtqueues'
      245 |   ret = virtio_create_virtqueues(vdev, 0, 1, vqname, callback);

-    virtio-pci/mmio: Features change to 64 bit in all virtio_dispatch and fix compile warning
    
    virtio/virtio-pci-legacy.c:399:48: warning: passing argument 2 of ‘pci_bus_write_io_dword’ makes integer from pointer without a cast [-Wint-con>
      399 |   pci_write_io_dword(vpdev->dev, vpdev->ioaddr + VIRTIO_PCI_GUEST_FEATURES,
          |                                                ^
          |                                                |
          |                                                void *
    /home/wyr/work/code/project/vela/nuttx/include/nuttx/pci/pci.h:559:65: note: expected ‘uintptr_t’ {aka ‘long unsigned int’} but argument is of >
      559 | int pci_bus_write_io_dword(FAR struct pci_bus_s *bus, uintptr_t where,
          |                                                       ~~~~~~~~~~^~~~~

-    virtio: adapt to the new OpenAMP
    
    Sync all the virtio api usage to sync with new OpenAMP

-    rpmsg: upgrade API passing on parameters with the upgrade of OpenAMP
    
    All the rpmsg transport layer and rpmsg services sync the API
    with new OpenAMP

-    openamp/libmetal.defs: fix OpenAMP upgrade err
    
    fix nuttx/openamp/libmetal/lib/cpu.h don't have <metal/processor/generic/cpu.h>
    error, let's use generic/cpu.h for all arch in NuttX

-    openamp: replace VIRTIO_DRIVER/DEVICE_ONLY to VIRTIO_DRIVER/DEVICE_SUPPORT
    
    New OpenAMP change the VIRTIO_DRIVER/DEVICE_ONLY to the
    VIRTIO_DRIVER_DEVCE_SUPPORT, so upgrade the NuttX OpenAMP config too

-    openamp: sync the openamp and libmetal to the newest commit
    
    Upgrade the openamp is a troublesome thing, so diretly sync the
    open-amp and libmetal the newest commit.
    
    We can update the commit and remove the patch when this patch has
    been merged in the OpenAMP repo.
    
## Impact
Rpmsg and VirtIO

## Testing
sim rpserver rpproxy and qemu armv8a

### CODE REVIEW

**Comment by CV-Bowen** [comment]
Split to two parts, part 1 is: https://github.com/apache/nuttx/pull/13919

**Comment by jerpelea** [comment]
@CV-Bowen master is usually a bad idea because it can introduce external bugs 
I think that we should stick to the latest stable release or a fixed commit

**Review by jerpelea — CHANGES_REQUESTED** [review]
please add commit messages explaining the change

**Comment by jerpelea** [comment]
1 mistake in an external project and we will have issues. I am still concerned with this release. Let's see what the others say!

**Comment by jerpelea** [comment]
I would like to hear other opinions

**Review by jerpelea — APPROVED** [review]
please add commit messages

**Comment by xiaoxiang781216** [comment]
> I would like to hear other opinions

the compromise proposal is pin to a specific commit number.

**Comment by jerpelea** [comment]
in my opinion this is something preferable since it is not influenced by new code

**Comment by hartmannathan** [comment]
> I would like to hear other opinions

Advantage of tracking releases only: Software we consume is in some kind of "known" state that is easy to document, e.g., "NuttX X.Y.Z uses FooBar 1.5" rather than some commit on a master branch.

Disadvantage of tracking releases only: We may need bug fixes that already exist on master but won't be released for a while. I think this is the same disadvantage that (non-rolling-release) OS distros have. They have to be synchronized to releases of all the included software, and if there are unreleased bugfixes that they need, then they need to backport the fix(es) and carry the patch(es). This is why some software has version numbers like "OpenSSH_9.2p1 Debian-2+deb12u2".

Advantage of tracking master: We get fixes (very close to immediately).

Disadvantage of tracking master: Potential for less stability. Even if the project is very careful about merges to master, they may be implementing new features and the changes can (potentially) disrupt us, even if they are good changes.

Hope this helps.

**Comment by hartmannathan** [comment]
Follow-up to my previous comment: I think the compromise proposal to pin to a specific commit is OK.

If we do that, then we should discuss (on mailing list) how often to advance to newer commits.

**Comment by xiaoxiang781216** [comment]
> Follow-up to my previous comment: I think the compromise proposal to pin to a specific commit is OK.
> 
> If we do that, then we should discuss (on mailing list) how often to advance to newer commits.

when a batch of nuttx specific change get merged, and pass our internal testing, we will update the pin to the new commit id. after the new official package which include all these fixes is released, we will switch back to track the release tag.

**Comment by xuxin930** [comment]
The CI check workflow checked the wrong file.
We will fix it in this commit https://github.com/apache/nuttx/pull/13959


image

File: openamp/libmetal.cmake [hunk 1]
```diff
@@ -22,9 +22,9 @@
 if(NOT EXISTS ${CMAKE_CURRENT_LIST_DIR}/libmetal)
   FetchContent_Declare(
     libmetal
-    DOWNLOAD_NAME "libmetal-v${OPENAMP_VERSION}.zip"
+    DOWNLOAD_NAME "libmetal-main.zip"
     DOWNLOAD_DIR ${CMAKE_CURRENT_LIST_DIR}
-    URL "https://github.com/OpenAMP/libmetal/archive/v${OPENAMP_VERSION}.zip"
+    URL "https://github.com/OpenAMP/libmetal/archive/"${LIBMETAL_COMMIT}".zip"
```
Comment (xuxin930) [inline openamp/libmetal.cmake:27]: Will quotes cause download failure? Should we remove the quotes here?
```suggestion
    URL "https://github.com/OpenAMP/libmetal/archive/${LIBMETAL_COMMIT}.zip"
```

File: openamp/open-amp.cmake [hunk 2]
```diff
@@ -22,9 +22,9 @@
 if(NOT EXISTS ${CMAKE_CURRENT_LIST_DIR}/open-amp)
   FetchContent_Declare(
     open-amp
-    DOWNLOAD_NAME "libopen-amp-v${OPENAMP_VERSION}.zip"
+    DOWNLOAD_NAME "libopen-amp-main.zip"
     DOWNLOAD_DIR ${CMAKE_CURRENT_LIST_DIR}
-    URL "https://github.com/OpenAMP/open-amp/archive/v${OPENAMP_VERSION}.zip"
+    URL "https://github.com/OpenAMP/open-amp/archive/"${OPENAMP_COMMIT}".zip"
```
Comment (xuxin930) [inline openamp/open-amp.cmake:27]:
```suggestion
    URL "https://github.com/OpenAMP/open-amp/archive/${OPENAMP_COMMIT}.zip"
```

File: openamp/libmetal.cmake (hunk 1 above)
Comment (CV-Bowen) [reply openamp/libmetal.cmake]: Done

File: openamp/open-amp.cmake (hunk 2 above)
Comment (CV-Bowen) [reply openamp/open-amp.cmake]: Done

**Comment by CV-Bowen** [comment]
@jerpelea @xiaoxiang781216 @hartmannathan @xuxin930 Done

**Review by jerpelea — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by cederom — APPROVED** [review]
I would also stick to a releases on external dependencies. But if you have a consensus on the master with existing fixes then okay. We should update dependency to a release as soon as it shows up :-)

Thank you @CV-Bowen :-)

**Comment by CV-Bowen** [comment]
@cederom Of course, we will pin to a release version after our PRs in OpenAMP community reach a conclusion.

**Pull request merged**